sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs

```diff
@@ -7,6 +7,14 @@ public partial class ApiKeyCredential
         public static implicit operator System.ClientModel.ApiKeyCredential (string key) { throw null; }
         public void Update(string key) { }
     }
+    public abstract partial class AsyncResultCollection<T> : System.ClientModel.ClientResult, System.Collections.Generic.IAsyncEnumerable<T>
+    {
+        protected internal AsyncResultCollection() : base (default(System.ClientModel.Primitives.PipelineResponse)) { }
+        protected internal AsyncResultCollection(System.ClientModel.Primitives.PipelineResponse response) : base (default(System.ClientModel.Primitives.PipelineResponse)) { }
+        public static System.ClientModel.AsyncResultCollection<System.BinaryData> Create(System.ClientModel.Primitives.PipelineResponse response, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public static System.ClientModel.AsyncResultCollection<TValue> Create<TValue>(System.Func<System.Threading.Tasks.Task<System.ClientModel.ClientResult>> getResultAsync, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) where TValue : System.ClientModel.Primitives.IJsonModel<TValue> { throw null; }
```

Comment:
@KrzysztofCwalina, what are your thoughts on putting a `Func<T>` in a public ClientModel API to handle postponing sending the request until the enumerable is foreached over?

Reply:
We have internal shared source files with `Func`s in them.
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Shared/PageableHelpers.cs

Reply:
hmm, what if we do the following: we will ship public Create method just for the protocol variant of the API, i.e. the one returning BinaryData events. This one does not need the delegate as for protocol methods we will not delay. For the delay scenario, can we have an abstraction/only, i.e. no Create factory method. The user (client library developer, like Travis) would have to inherit and override a method that get the response.

Reply:
That should work.  Are you happy with the client author (i.e. Travis) layering their implementation over the BinaryData implementation and deserializing from BinaryData?  If they don't use this approach, they would need to reimplement the SSE parser/reader types internal to the client library.

I can mock up an example if needed.

Reply:
Actually, I'll add a unit test showing the proposal.

Reply:
I believe this is addressed, but please let me know if there is outstanding feedback here.